tests/python/test_slowmo_fsdp.py

```diff
@@ -0,0 +1,382 @@
+        # to differ from `slowmo_freq` to check it independently from
+        # the momentum's update. Basically, parameter averaging now will happen
+        # every 3rd step and momentum step every 6th.
+        slowmo_optim.averager.period = 3
+
+        averager2 = averagers.PeriodicModelAverager(
+            period=3, process_group=dist.distributed_c10d._get_default_group()
+        )
+        base_optimizer = torch.optim.SGD(fsdp_net.parameters(), lr=1e-2)
+
+        for _ in range(4):
+            self._train_step(inpt, fsdp_net, base_optimizer)
+            self._train_step(inpt, fsdp_net_slowmo, slowmo_optim)
+            averager2.average_parameters(list(fsdp_net.parameters()))
+
+        for slowmo_params, net_params in zip(
+            fsdp_net_slowmo.parameters(), fsdp_net.parameters()
+        ):
+            self.assertEqual(slowmo_params, net_params)
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize("sharding_strategy", [ShardingStrategy.NO_SHARD])
+    def test_slowmo_optimizer_momentum_step(
+        self, sharding_strategy: Optional[ShardingStrategy]
+    ):
+        net = torch.nn.Linear(2, 1)
+
+        fsdp_net = FSDP(
+            copy.deepcopy(net),
+            device_id=torch.cuda.current_device(),
+            sharding_strategy=sharding_strategy,
+        ).to(self.rank)
+        fsdp_net_slowmo = FSDP(
+            copy.deepcopy(net),
+            device_id=torch.cuda.current_device(),
+            sharding_strategy=sharding_strategy,
+        ).to(self.rank)
+
+        cur_subgroup = dist.new_group(ranks=[self.rank])
+        slowMoState = slow_momentum_comm.SlowMoState(cur_subgroup)
+        fsdp_net.register_comm_hook(slowMoState, slow_momentum_comm.slowmo_hook)
+        fsdp_net_slowmo.register_comm_hook(slowMoState, slow_momentum_comm.slowmo_hook)
+        inpt = torch.tensor([(self.rank + 1)] * 2).float()
+        inpt.to(self.rank)  # type: ignore[call-overload]
+
+        slowmo_optim = slow_momentum_optimizer.SlowMomentumOptimizer(
+            base_optim=torch.optim.SGD(fsdp_net_slowmo.parameters(), lr=1e-2),
+            slowmo_freq=1,
+            slowmo_factor=0.5,
+            slowmo_lr=0.1,
+        )
+
+        averager2 = averagers.PeriodicModelAverager(
+            period=1, process_group=dist.distributed_c10d._get_default_group()
+        )
+        base_optimizer = torch.optim.SGD(fsdp_net.parameters(), lr=1e-2)
+
+        for _, v in slowmo_optim.state.items():
+            initial_prev_params = copy.deepcopy(v["prev_param"])
+            initial_slow_momentum_buffer = copy.deepcopy(v["slow_momentum"])
+
+        for _ in range(1):
+            self._train_step(inpt, fsdp_net, base_optimizer)
+            self._train_step(inpt, fsdp_net_slowmo, slowmo_optim)
+            averager2.average_parameters(list(fsdp_net.parameters()))
+
+        # parameters before slow momentum update and after averaging
+        # are in `fsdp_net.params[0]`
```

Comment:
Would it be correct to say that this test requires there to only be a top-level FSDP wrapper (without nesting)?

Reply:
@awgu this test - yes. Would it be better to also test a nested wrapping?

Reply:
Nope. I do not think there is a need to test a nested wrapping. I just wanted to check. Thanks!